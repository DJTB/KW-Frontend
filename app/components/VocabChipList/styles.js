import styled, { injectGlobal } from 'styled-components';

import { gutter } from 'shared/styles/layout';
import { epsilon } from 'shared/styles/typography';

import { white, grey } from 'shared/styles/colors';
import { borderRadius } from 'shared/styles/sizing';
import { resetList } from 'shared/styles/utils';

/* eslint no-unused-expressions: 0 */
injectGlobal`
  /* VocabChip tooltip styles */
  .vocab-tip.vocab-tip {
    ${epsilon}
    padding: .4em .6em .5em;
    border-radius: ${borderRadius};
  }

  .vocab-tip ul {
    ${resetList}
    color: ${white};
  }

  .vocab-tip li {
    display: flex;
    align-items: center;
    justify-content: space-between;

    span {
      flex: 0 1 auto;
      padding: .2em;
    }

    /* left column category "JA", "RC" etc */
    span:first-child {
      font-size: .95em;
      color: ${grey};
      margin-right: .2em;
    }
  }
`;

export const Ul = styled.ul`
  ${resetList}
  ${gutter()}
`;
