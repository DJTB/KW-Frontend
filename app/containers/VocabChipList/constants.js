export const DEFAULT_ACTION = 'app/VocabChipList/DEFAULT_ACTION';
