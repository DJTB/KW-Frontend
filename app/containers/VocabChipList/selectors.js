import { createSelector } from 'reselect';

// Direct selector to the vocabChipList state domain
const selectVocabChipListDomain = () => (state) => state.vocabChipList;

// Default selector used by VocabChipList
const makeSelectVocabChipList = () => createSelector(
  selectVocabChipListDomain(),
  (substate) => substate
);

// Other specific selectors

export default selectVocabChipListDomain;
export {
  makeSelectVocabChipList,
};
