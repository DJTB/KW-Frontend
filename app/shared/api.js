import urljoin from 'url-join';
import { get, put, post, patch, del } from 'utils/request';

const BASE_URL = '//localhost:8000'; // FIXME: change for production
const API_BASE = urljoin(BASE_URL, 'api', 'v1');

//-----------------------------------------------------------------------------
//  AUTHORIZATION
//-----------------------------------------------------------------------------
const authUrl = urljoin(API_BASE, 'auth');
const userLoginUrl = urljoin(authUrl, 'login');
const userCredentialsUrl = urljoin(authUrl, 'me'); // GET / PATCH
const registerUrl = urljoin(authUrl, 'register'); // POST
const activateUrl = urljoin(authUrl, 'activate'); // POST
const usernameUrl = urljoin(authUrl, 'username'); // POST
const passwordUrl = urljoin(authUrl, 'password'); // POST
const resetPasswordUrl = urljoin(passwordUrl, 'reset'); // POST
const confirmPasswordUrl = urljoin(resetPasswordUrl, 'confirm'); // POST

//-----------------------------------------------------------------------------
//  USER
//-----------------------------------------------------------------------------
const userUrl = urljoin(API_BASE, 'user'); // GET all users (if admin, else 'me')
const userProfileUrl = urljoin(userUrl, 'me'); // GET user profile
const userSrsUrl = urljoin(userUrl, 'srs'); // POST to get review count
const userSyncUrl = urljoin(userUrl, 'sync'); // POST to sync with WK

//-----------------------------------------------------------------------------
//  REVIEWS
//-----------------------------------------------------------------------------
const reviewsUrl = urljoin(API_BASE, 'review'); // GET all ready reviews
const criticalReviewsUrl = urljoin(reviewsUrl, 'critical'); // GET critical
const currentReviewsUrl = urljoin(reviewsUrl, 'current'); // GET current review queue
const currentLessonsUrl = urljoin(reviewsUrl, 'lesson'); // GET current lesson queue
const reviewEntryUrl = (id) => urljoin(reviewsUrl, id); // GET single
const reviewCorrectUrl = (id) => urljoin(reviewEntryUrl(id), 'correct'); // POST correct answer
const reviewIncorrectUrl = (id) => urljoin(reviewEntryUrl(id), 'incorrect'); // POST incorrect answer
const lockReviewUrl = (id) => urljoin(reviewEntryUrl(id), 'hide'); // POST
const unlockReviewUrl = (id) => urljoin(reviewEntryUrl(id), 'unhide'); // POST

//-----------------------------------------------------------------------------
//  SYNONYMS
//-----------------------------------------------------------------------------
const synonymUrl = urljoin(API_BASE, 'synonym'); // POST add, GET get all
const synonymEntryUrl = (id) => urljoin(synonymUrl, id); // DELETE remove, GET one

//-----------------------------------------------------------------------------
//  VOCABULARY
//-----------------------------------------------------------------------------
const vocabularyUrl = urljoin(API_BASE, 'vocabulary'); // GET everything!
const vocabularyEntryUrl = (id) => urljoin(vocabularyUrl, id); // GET one
const readingUrl = urljoin(API_BASE, 'reading'); // GET all
const readingEntryUrl = (id) => urljoin(readingUrl, id); // GET one
const levelsUrl = urljoin(API_BASE, 'level'); // GET all
const levelEntryUrl = (id) => urljoin(levelsUrl, id); // GET one
const lockLevelUrl = (id) => urljoin(levelEntryUrl(id), 'lock'); // POST lock
const unlockLevelUrl = (id) => urljoin(levelEntryUrl(id), 'unlock'); // POST unlock

//-----------------------------------------------------------------------------
//  GENERAL
//-----------------------------------------------------------------------------
const faqUrl = urljoin(API_BASE, 'faq'); // GET all, POST create
const faqEntryUrl = (id) => urljoin(faqUrl, id); // GET, PUT update, PATCH partial update, DELETE
const announcementUrl = urljoin(API_BASE, 'announcement'); // GET all, POST create
const announcementEntryUrl = (id) => urljoin(announcementUrl, id); // GET one, PUT update, PATCH  partial update, DELETE
const contactUrl = urljoin(API_BASE, 'contact');

//-----------------------------------------------------------------------------
//  EXTERNAL
//-----------------------------------------------------------------------------
export const createJishoApiUrl = (keyword) => `//jisho.org/api/v1/search/words?keyword=${keyword}`;
export const createJishoUrl = (keyword) => `//jisho.org/search/${keyword}`;


//-----------------------------------------------------------------------------
//  AUTHORIZATION
//-----------------------------------------------------------------------------
export const getUserAuth = ({ id, username, email } = {}) => get(userCredentialsUrl, { id, username, email });
export const loginUser = ({ username, password } = {}) => post(userLoginUrl, { username, password });
export const updateUserAuth = ({ id, username, email } = {}) => patch(userCredentialsUrl, { id, username, email });
export const registerUser = ({ email, username, password, apiKey } = {}) =>
  post(registerUrl, { email, username, password, api_key: apiKey });

export const activateUser = ({ uid } = {}) => post(activateUrl, { uid });
export const changeUsername = ({ username } = {}) => post(usernameUrl, { username });
export const changePassword = ({ password } = {}) => post(passwordUrl, { password });
export const resetPassword = ({ email } = {}) => post(resetPasswordUrl, { email }); // send password reset email.
export const confirmPassword = () => post(confirmPasswordUrl); // finish reset password process

//-----------------------------------------------------------------------------
//  USER
//-----------------------------------------------------------------------------
export const getUsers = () => get(userUrl);
export const getUserProfile = () => get(userProfileUrl);
export const syncKw = () => post(userSrsUrl);
// true to force ALL users to sync with WK
export const syncWk = ({ fullSync = false } = {}) => post(userSyncUrl, { full_sync: fullSync });

//-----------------------------------------------------------------------------
//  REVIEWS
//-----------------------------------------------------------------------------
export const getReviews = ({
  id,
  offset,
  limit,
  meaningContains,
  srsLevel,
  srsLevelLt,
  srsLevelGt,
} = {}) => get(
  reviewsUrl,
  {
    level: id,
    offset,
    limit,
    meaning_contains: meaningContains,
    srs_level: srsLevel,
    srs_level_lt: srsLevelLt,
    srs_level_gt: srsLevelGt,
  }
);

export const getCriticalReviews = ({ offset, limit } = {}) => get(criticalReviewsUrl, { offset, limit });
export const getCurrentReviews = ({ offset, limit } = {}) => get(currentReviewsUrl, { offset, limit });
export const getCurrentLessons = ({ offset, limit } = {}) => get(currentLessonsUrl, { offset, limit });

export const getReviewEntry = ({ id }) => get(reviewEntryUrl(id));
export const recordReview = ({ id, isCorrect, previouslyIncorrect }) => {
  const body = {
    user_specific_id: id,
    user_correct: isCorrect,
    wrong_before: previouslyIncorrect,
  };
  return isCorrect ? post(reviewCorrectUrl(id), body) : post(reviewIncorrectUrl(id), body);
};

export const lockReview = ({ id }) => post(lockReviewUrl(id));
export const unlockReview = ({ id }) => post(unlockReviewUrl(id));
export const saveReviewNotes = ({ id, notes }) => patch(reviewEntryUrl(id), { notes });

//-----------------------------------------------------------------------------
//  SYNONYMS
//-----------------------------------------------------------------------------
export const addSynonym = ({ reviewId, character, kana } = {}) => post(
  synonymUrl,
  { review: reviewId, character, kana },
);
export const removeSynonym = ({ id }) => del(synonymEntryUrl(id));

//-----------------------------------------------------------------------------
//  VOCABULARY
//-----------------------------------------------------------------------------
export const getVocabulary = ({
  id: level,
  offset,
  limit,
  meaningContains,
  readingsKanaContains,
  readingsCharacterContains,
} = {}) => get(
  vocabularyUrl,
  {
    level,
    offset,
    limit,
    meaning_contains: meaningContains,
    readings_kana_contains: readingsKanaContains,
    readings_character_contains: readingsCharacterContains,
  }
);

export const getLevels = () => get(levelsUrl);

export const getVocabularyEntry = ({ id }) => get(vocabularyEntryUrl(id));

export const getReadings = ({ offset, limit } = {}) => get(readingUrl, { offset, limit });
export const getReadingEntry = ({ id }) => get(readingEntryUrl(id));

export const lockLevel = ({ id }) => post(lockLevelUrl(id));
export const unlockLevel = ({ id }) => post(unlockLevelUrl(id));

//-----------------------------------------------------------------------------
//  GENERAL
//-----------------------------------------------------------------------------
export const getFaqs = () => get(faqUrl);
export const getFaq = ({ id }) => get(faqEntryUrl(id));
export const addFaq = ({ question, answer } = {}) => post(faqUrl, { question, answer });
export const updateFaq = ({ question, answer } = {}) => put(faqUrl, { question, answer });

export const getAnnouncements = () => get(announcementUrl);
export const getAnnouncement = ({ id }) => get(announcementEntryUrl(id));
export const addAnnouncement = ({ title, body } = {}) => post(announcementUrl, { title, body });
export const updateAnnouncement = ({ title, body } = {}) => put(announcementUrl, { title, body });

export const sendContactMessage = ({ name, email, body } = {}) => post(contactUrl, { name, email, body });
