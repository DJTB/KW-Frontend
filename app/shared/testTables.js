import { random } from 'lodash';

/**
 * These all assume that the data has been serialized to preferred format in the front end already!
 * This is not the API response shape.
 */

export const levels = [
  {
    level: 1,
    unlocked: false,
    count: 43,
  },
  {
    level: 2,
    unlocked: true,
    count: 90,
  },
  {
    level: 3,
    unlocked: true,
    count: 69,
  },
  {
    level: 4,
    unlocked: false,
    count: 105,
  },
  {
    level: 5,
    unlocked: true,
    count: 126,
  },
  {
    level: 6,
    unlocked: true,
    count: 118,
  },
  {
    level: 7,
    unlocked: true,
    count: 101,
  },
  {
    level: 8,
    unlocked: false,
    count: 136,
  },
  {
    level: 9,
    unlocked: false,
    count: 119,
  },
  {
    level: 10,
    unlocked: false,
    count: 118,
  },
  {
    level: 11,
    unlocked: false,
    count: 126,
  },
  {
    level: 12,
    unlocked: false,
    count: 126,
  },
  {
    level: 13,
    unlocked: false,
    count: 115,
  },
  {
    level: 14,
    unlocked: false,
    count: 116,
  },
  {
    level: 15,
    unlocked: false,
    count: 100,
  },
  {
    level: 16,
    unlocked: false,
    count: 117,
  },
  {
    level: 17,
    unlocked: false,
    count: 124,
  },
  {
    level: 18,
    unlocked: false,
    count: 135,
  },
  {
    level: 19,
    unlocked: false,
    count: 106,
  },
  {
    level: 20,
    unlocked: false,
    count: 114,
  },
  {
    level: 21,
    unlocked: false,
    count: 106,
  },
  {
    level: 22,
    unlocked: false,
    count: 115,
  },
  {
    level: 23,
    unlocked: false,
    count: 97,
  },
  {
    level: 24,
    unlocked: false,
    count: 120,
  },
  {
    level: 25,
    unlocked: false,
    count: 100,
  },
  {
    level: 26,
    unlocked: false,
    count: 124,
  },
  {
    level: 27,
    unlocked: false,
    count: 106,
  },
  {
    level: 28,
    unlocked: false,
    count: 116,
  },
  {
    level: 29,
    unlocked: false,
    count: 107,
  },
  {
    level: 30,
    unlocked: false,
    count: 103,
  },
  {
    level: 31,
    unlocked: false,
    count: 113,
  },
  {
    level: 32,
    unlocked: false,
    count: 111,
  },
  {
    level: 33,
    unlocked: false,
    count: 103,
  },
  {
    level: 34,
    unlocked: false,
    count: 124,
  },
  {
    level: 35,
    unlocked: false,
    count: 100,
  },
  {
    level: 36,
    unlocked: false,
    count: 98,
  },
  {
    level: 37,
    unlocked: false,
    count: 112,
  },
  {
    level: 38,
    unlocked: false,
    count: 100,
  },
  {
    level: 39,
    unlocked: false,
    count: 101,
  },
  {
    level: 40,
    unlocked: false,
    count: 116,
  },
  {
    level: 41,
    unlocked: false,
    count: 107,
  },
  {
    level: 42,
    unlocked: false,
    count: 96,
  },
  {
    level: 43,
    unlocked: false,
    count: 108,
  },
  {
    level: 44,
    unlocked: false,
    count: 103,
  },
  {
    level: 45,
    unlocked: false,
    count: 96,
  },
  {
    level: 46,
    unlocked: false,
    count: 103,
  },
  {
    level: 47,
    unlocked: false,
    count: 94,
  },
  {
    level: 48,
    unlocked: false,
    count: 104,
  },
  {
    level: 49,
    unlocked: false,
    count: 104,
  },
  {
    level: 50,
    unlocked: false,
    count: 87,
  },
  {
    level: 51,
    unlocked: false,
    count: 75,
  },
  {
    level: 52,
    unlocked: false,
    count: 99,
  },
  {
    level: 53,
    unlocked: false,
    count: 103,
  },
  {
    level: 54,
    unlocked: false,
    count: 112,
  },
  {
    level: 55,
    unlocked: false,
    count: 76,
  },
  {
    level: 56,
    unlocked: false,
    count: 88,
  },
  {
    level: 57,
    unlocked: false,
    count: 93,
  },
  {
    level: 58,
    unlocked: false,
    count: 81,
  },
  {
    level: 59,
    unlocked: false,
    count: 79,
  },
  {
    level: 60,
    unlocked: false,
    count: 65,
  },
];

export const vocabs = [
  {
    id: 553,
    history: {
      correct: random(10),
      incorrect: random(10),
    },
    session: {
      correct: random(10),
      incorrect: random(10),
      streak: random(11),
    },
    meanings: ['before long', 'soon'],
    readings: [
      {
        id: 582,
        character: '近々',
        kana: ['ちかぢか'],
        level: 7,
        tags: ['Adverbial Noun'],
        sentenceEn: 'When I saw them on a date, I asked them, “Do I hear wedding bells?”',
        sentenceJa: '彼らがデートをしているのを見かけて近々結婚するのってきいたんだ',
        jlpt: 'JLPT N2',
        common: true,
      },
      {
        id: 583,
        character: '近々',
        kana: ['きんきん'],
        level: 7,
        tags: ['Adverbial Noun'],
        sentenceEn: 'When I saw them on a date, I asked them, “Do I hear wedding bells?”',
        sentenceJa: '彼らがデートをしているのを見かけて近々結婚するのってきいたんだ',
        jlpt: 'JLPT N2',
        common: true,
      },
      {
        id: 584,
        character: '近々',
        kana: ['ちかじか'],
        level: 7,
        tags: ['Adverbial Noun'],
        sentenceEn: 'When I saw them on a date, I asked them, “Do I hear wedding bells?”',
        sentenceJa: '彼らがデートをしているのを見かけて近々結婚するのってきいたんだ',
        jlpt: 'JLPT N2',
        common: true,
      },
      {
        id: 783,
        character: '軈て',
        kana: ['やがて'],
        level: 61,
        tags: ['Adverb'],
        sentenceEn: 'By and by he forgot that affair',
        sentenceJa: 'やがて彼はその事件のことを忘れた',
        jlpt: 'JLPT N2',
        common: true,
      },
    ],
    synonyms: [
      {
        id: 5582,
        character: '間もなく',
        kana: ['まもなく'],
        level: 22,
        tags: ['Adverb'],
        sentenceEn: 'Coming up, Jiyuugaoka Station.',
        sentenceJa: '間もなく、自由が丘駅',
        jlpt: 'JLPT N2',
        common: true,
      },
    ],
  },
  {
    id: 176,
    history: {
      correct: random(10),
      incorrect: random(10),
    },
    session: {
      correct: random(10),
      incorrect: random(10),
      streak: random(11),
    },
    meanings: ['young lady', 'little girl', 'maiden', 'daughter', 'girl'],
    readings: [
      {
        id: 188,
        character: '少女',
        kana: ['しょうじょ'],
        level: 3,
        tags: ['Noun'],
        sentenceEn: "I am touched by the girl's acute sensitivity.",
        sentenceJa: '少女の鋭い感性に触れている',
        jlpt: 'JLPT N3',
        common: true,
      },
      {
        id: 189,
        character: '少女',
        kana: ['おとめ'],
        level: 3,
        tags: ['Noun'],
        sentenceEn: "I am touched by the girl's acute sensitivity.",
        sentenceJa: '少女の鋭い感性に触れている',
        jlpt: 'JLPT N3',
        common: true,
      },
    ],
    synonyms: [],
  },
  {
    id: 280,
    history: {
      correct: random(10),
      incorrect: random(10),
    },
    session: {
      correct: random(10),
      incorrect: random(10),
      streak: random(11),
    },
    meanings: ['young boy', 'juvenile', 'youth', 'boy'],
    readings: [
      {
        id: 298,
        character: '少年',
        kana: ['しょうねん'],
        level: 4,
        tags: ['Noun'],
        sentenceEn: 'The shy boy murmured his name.',
        sentenceJa: 'その内気な少年はぼそぼそと自分の名前をつぶやいた全ての少年少女はを知っています',
        jlpt: 'JLPT N3',
        common: true,
      },
    ],
    synonyms: [],
  },
  {
    id: 424,
    history: {
      correct: random(10),
      incorrect: random(10),
    },
    session: {
      correct: random(10),
      incorrect: random(10),
      streak: random(11),
    },
    meanings: [
      'whole life',
      'entire life',
      'a lifetime',
      'all through life',
      'one existence',
      'a generation',
      'an age',
    ],
    readings: [
      {
        id: 449,
        character: '一生',
        kana: ['いっしょう'],
        level: 5,
        tags: ['Noun', 'No adjective'],
        sentenceEn: 'She worked hard.',
        sentenceJa: '彼女は一生懸命働いた',
        jlpt: 'JLPT N3',
        common: true,
      },
    ],
    synonyms: [],
  },
  {
    id: 502,
    history: {
      correct: random(10),
      incorrect: random(10),
    },
    session: {
      correct: random(10),
      incorrect: random(10),
      streak: random(11),
    },
    meanings: ['honest'],
    readings: [
      {
        id: 529,
        character: '正直',
        kana: ['しょうじき'],
        level: 6,
        tags: ['Noun', 'Na adjective'],
        sentenceEn: 'Him, honest? What a joke!',
        sentenceJa: '正直が聞いて呆れるよ',
        jlpt: 'JLPT N3',
        common: true,
      },
    ],
    synonyms: [],
  },
];

export const readings = {
  single: [
    {
      id: 351,
      character: '売り上げ',
      kana: ['うりあげ'],
      level: 5,
      tags: ['Noun'],
      sentenceEn: 'Sales are down this month.',
      sentenceJa: '今月は売上が落ちた',
      jlpt: 'JLPT N2',
      common: true,
    },
  ],
  sameCharacters: [
    {
      id: 582,
      character: '近々',
      kana: ['ちかぢか'],
      level: 7,
      tags: ['Noun'],
      sentenceEn: 'When I saw them on a date, I asked them, “Do I hear wedding bells?”',
      sentenceJa: '彼らがデートをしているのを見かけて近々結婚するのってきいたんだ',
      jlpt: 'JLPT N2',
      common: true,
    },
    {
      id: 583,
      character: '近々',
      kana: ['きんきん'],
      level: 7,
      tags: ['Noun'],
      furigana: '0:きんきん',
      sentenceEn: 'When I saw them on a date, I asked them, “Do I hear wedding bells?”',
      sentenceJa: '彼らがデートをしているのを見かけて近々結婚するのってきいたんだ',
      jlpt: 'JLPT N2',
      common: true,
    },
    {
      id: 584,
      character: '近々',
      kana: ['ちかじか'],
      level: 7,
      tags: ['Noun'],
      sentenceEn: 'When I saw them on a date, I asked them, “Do I hear wedding bells?”',
      sentenceJa: '彼らがデートをしているのを見かけて近々結婚するのってきいたんだ',
      jlpt: 'JLPT N2',
      common: true,
    },
  ],
  suffix: [
    {
      id: 2476,
      character: '各〜',
      kana: ['かく'],
      level: 22,
      tags: ['Prefix'],
      sentenceEn: 'First, choose a representative for each team.',
      sentenceJa: 'まず始めに、各チームの代表者を決めて下さい',
      jlpt: null,
      common: null,
    },
  ],
  prefix: [
    {
      id: 6337,
      character: '〜漬け',
      kana: ['づけ'],
      level: 59,
      tags: ['Suffix'],
      sentenceEn: 'These are lightly pickled cucumbers.',
      sentenceJa: '胡瓜の浅漬けならありますよ',
      jlpt: null,
      common: null,
    },
    {
      id: 130,
      character: '〜才',
      kana: ['さい'],
      level: 2,
      tags: ['Suffix'],
      sentenceEn: 'I am five years old.',
      sentenceJa: 'わたしは５才です',
      jlpt: null,
      common: null,
    },
  ],
  mixed: [
    {
      id: 933,
      character: 'ハート形',
      kana: ['ハートがた'],
      level: 9,
      tags: ['Noun'],
      sentenceEn: 'When flamingos kiss, their heads and necks form a heart shape.',
      sentenceJa: 'フラミンゴがキスをする時、二羽の頭と首がハート形を描く',
      jlpt: null,
      common: false,
    },
    {
      id: 30,
      character: 'フランス人',
      kana: ['ふらんすじん'],
      level: 1,
      tags: ['Noun'],
      sentenceEn: 'He is French.',
      sentenceJa: '彼はフランス人です',
      jlpt: null,
      common: false,
    },
    {
      id: 3190,
      character: '迷惑メール',
      kana: ['めいわくメール'],
      level: 29,
      tags: ['Noun'],
      sentenceEn:
        'He sent about 2.2 billion spam emails over a period of around one and a half years.',
      sentenceJa: '約１年半で約億通の迷惑メールを送りました',
      jlpt: null,
      common: false,
    },
    {
      id: 6193,
      character: '烏龍茶',
      kana: ['ウーロンちゃ'],
      level: 57,
      tags: ['Noun'],
      sentenceEn: 'Oolong tea without ice, please.',
      sentenceJa: '烏龍茶を氷抜きでください',
      jlpt: null,
      common: null,
    },
    {
      id: 4875,
      character: '缶ビール',
      kana: ['かんビール'],
      level: 44,
      tags: ['Noun'],
      sentenceEn: 'Oh, this canned beer is way too good for me.',
      sentenceJa: 'ああ、この缶ビールはマジでうますぎる',
      jlpt: null,
      common: false,
    },
  ],
  onlyKana: [
    {
      id: 9000,
      common: true,
      jlpt: 'JLPT N4',
      character: 'そろそろ',
      kana: ['そろそろ'],
      tags: ['Adverb', "Adverb taking the 'to' particle"],
      sentenceJa: '「あの、かりんさん...、そろそろお開きにしませんか？」「駄目、勝ち逃げ禁止！今度こそボクが勝つの！」',
      sentenceEn:
        "“Er, Karin ..., shouldn't we call it a night soon?” “No! No quitting while ahead! Next time I'll win for sure!”",
    },
    {
      id: 9001,
      common: false,
      jlpt: null,
      character: 'マンコ',
      kana: ['マンコ'],
      tags: ['Noun'],
      sentenceJa: 'マンコいいな〜',
      sentenceEn: 'What a great pussy.',
    },
  ],
  partialKanaSentence: [
    {
      id: 8800,
      common: true,
      jlpt: 'JLPT N4',
      character: '付く',
      kana: ['つく'],
      tags: ['Transitive Verb'],
      sentenceJa: 'この「オンボード」というのは平たく言えば「グラフィックカードがついていない」ものだ',
      sentenceEn:
        'Frankly speaking, this “on board” means “doesn’t have a graphics card”.',
    },
  ],
};
