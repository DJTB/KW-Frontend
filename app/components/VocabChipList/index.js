import React from 'react';
import PropTypes from 'prop-types';
import cuid from 'cuid';
import titleCase from 'voca/title_case';

import ReactTooltip from 'react-tooltip';

import VocabChip from 'components/VocabChip';
import WithTooltip from 'utils/WithTooltip';
import calculateCorrectnessPercentage from 'utils/calculateCorrectnessPercentage';
import * as COLORS from 'shared/styles/colors';

import { Ul } from './styles';

VocabChipList.propTypes = {
  items: PropTypes.array.isRequired,
  color: PropTypes.oneOf(Object.keys(COLORS)),
};

VocabChipList.defaultProps = {
  color: 'purple',
};

// `.vocab-tip` tooltip styles are injected in `globalStyles.js`
const VocabChipWithToolTip = WithTooltip(VocabChip);

const generateToolTip = (meaning, kana, correctPercent) => `
  <ul>
    <li>
      <span>EN</span>
      <span>${meaning}</span>
    <li>
      <span>JA </span>
      <span lang="ja">${kana}</span>
    </li>
    <li>
      <span>RC</span>
      <span>${correctPercent}%</span>
    </li>
  </ul>
`;

function VocabChipList({ items, color }) {
  return (
    <div>
      <ReactTooltip
        id="vocabChipTip"
        className="vocab-tip"
        html
      />
      <Ul>
        {/* FIXME: memoize */}
        {items.map(({ id, meanings, readings, history, session }) => {
          const meaning = titleCase(meanings[0]);
          const { kana, character } = readings[0];
          const correctPercent = calculateCorrectnessPercentage(history, session);

          const tooltipText = generateToolTip(meaning, kana, correctPercent);
          return (
            <VocabChipWithToolTip
              key={cuid()}
              id={id}
              color={color}
              character={character}
              data-for="vocabChipTip"
              data-tip={tooltipText}
              data-place="bottom"
            />
          );
        })}
      </Ul>
    </div>
  );
}

export default VocabChipList;
