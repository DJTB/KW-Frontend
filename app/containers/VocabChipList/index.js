import React from 'react';
import PropTypes from 'prop-types';
import cuid from 'cuid';

import ReactTooltip from 'react-tooltip';

import Ul from 'base/Ul';
import VocabChip from 'components/VocabChip';
import WithTooltip from 'containers/WithTooltip';
import * as COLORS from 'shared/styles/colors';

VocabChipList.propTypes = {
  items: PropTypes.array.isRequired,
  color: PropTypes.oneOf(Object.keys(COLORS)),
};

VocabChipList.defaultProps = {
  color: 'purple',
};

// FIXME: operate in item selector instead
const generateToolTip = ({ meaning, kana, correctPercent }) => `
  <ul>
    <li>
      <span>EN</span>
      <span>${meaning}</li>
    <li>
      <span>JA</span>
      <span lang="ja">${kana}
    </li>
    <li>
      <span>RC</span>
      <span>${correctPercent}%
    </li>
  </ul>
`;

// SELECTOR map over items!
// import titleCase from 'voca/title_case';
// import calculateCorrectnessPercentage from 'utils/calculateCorrectnessPercentage';
// meaning = titleCase(item.meanings[0]);
// { kana, character } = item.readings[0];
// correctPercent = calculateCorrectnessPercentage(item.history, item.session);
// tooltipText = generateToolTip(item)

// `.vocab-tip` tooltip styles are injected in `globalStyles.js`
const VocabChipWithToolTip = WithTooltip(VocabChip);

function VocabChipList({ items, color, ...props }) {
  return (
    <div>
      <ReactTooltip
        id="vocabChipTip"
        className="vocab-tip"
        html
      />
      <Ul plainList {...props}>
        {items.map((item) => {
          const tooltipText = generateToolTip(item); // FIXME: map onto item in memoized item selector instead
          return (
            <VocabChipWithToolTip
              key={cuid()}
              id={item.id}
              color={color}
              character={item.character}
              data-for="vocabChipTip"
              data-tip={tooltipText}
            />
          );
        })}
      </Ul>
    </div>
  );
}

export default VocabChipList;
