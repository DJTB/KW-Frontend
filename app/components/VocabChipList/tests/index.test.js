import React from 'react';
import { shallow } from 'enzyme';

import { vocabs } from 'shared/testTables';
import VocabChipList from '../index';

describe('<VocabChipList />', () => {
  const items = vocabs.slice(0, 3).map((vocab) => ({
    ...vocab,
    history: {
      correct: 4,
      incorrect: 2,
    },
    session: {
      correct: 1,
      incorrect: 1,
    },
  }));

  it('should match baseline snapshot', () => {
    const renderedComponent = shallow(
      <VocabChipList
        items={items}
      />
    );
    expect(renderedComponent).toMatchSnapshot();
  });

  it('should adopt a color prop', () => {
    const renderedComponent = shallow(
      <VocabChipList
        items={items}
        color="green"
      />
    );
    expect(renderedComponent).toMatchSnapshot();
  });
});
