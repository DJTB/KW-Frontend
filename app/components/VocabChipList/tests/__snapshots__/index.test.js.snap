// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VocabChipList /> should adopt a color prop 1`] = `
<div>
  <ReactTooltip
    className="vocab-tip"
    html={true}
    id="vocabChipTip"
    insecure={true}
    resizeHide={true}
    wrapper="div"
  />
  <styled.ul>
    <defaultProps(VocabChip)
      character="近々"
      color="green"
      data-for="vocabChipTip"
      data-place="bottom"
      data-tip="
      <ul>
        <li>
          <span>EN</span>
          <span>Before Long</span>
        <li>
          <span>JA </span>
          <span lang=\\"ja\\">ちかぢか</span>
        </li>
        <li>
          <span>RC</span>
          <span>62%</span>
        </li>
      </ul>
    "
      id={553}
    />
    <defaultProps(VocabChip)
      character="少女"
      color="green"
      data-for="vocabChipTip"
      data-place="bottom"
      data-tip="
      <ul>
        <li>
          <span>EN</span>
          <span>Young Lady</span>
        <li>
          <span>JA </span>
          <span lang=\\"ja\\">しょうじょ</span>
        </li>
        <li>
          <span>RC</span>
          <span>62%</span>
        </li>
      </ul>
    "
      id={176}
    />
    <defaultProps(VocabChip)
      character="少年"
      color="green"
      data-for="vocabChipTip"
      data-place="bottom"
      data-tip="
      <ul>
        <li>
          <span>EN</span>
          <span>Young Boy</span>
        <li>
          <span>JA </span>
          <span lang=\\"ja\\">しょうねん</span>
        </li>
        <li>
          <span>RC</span>
          <span>62%</span>
        </li>
      </ul>
    "
      id={280}
    />
  </styled.ul>
</div>
`;

exports[`<VocabChipList /> should match baseline snapshot 1`] = `
<div>
  <ReactTooltip
    className="vocab-tip"
    html={true}
    id="vocabChipTip"
    insecure={true}
    resizeHide={true}
    wrapper="div"
  />
  <styled.ul>
    <defaultProps(VocabChip)
      character="近々"
      color="purple"
      data-for="vocabChipTip"
      data-place="bottom"
      data-tip="
      <ul>
        <li>
          <span>EN</span>
          <span>Before Long</span>
        <li>
          <span>JA </span>
          <span lang=\\"ja\\">ちかぢか</span>
        </li>
        <li>
          <span>RC</span>
          <span>62%</span>
        </li>
      </ul>
    "
      id={553}
    />
    <defaultProps(VocabChip)
      character="少女"
      color="purple"
      data-for="vocabChipTip"
      data-place="bottom"
      data-tip="
      <ul>
        <li>
          <span>EN</span>
          <span>Young Lady</span>
        <li>
          <span>JA </span>
          <span lang=\\"ja\\">しょうじょ</span>
        </li>
        <li>
          <span>RC</span>
          <span>62%</span>
        </li>
      </ul>
    "
      id={176}
    />
    <defaultProps(VocabChip)
      character="少年"
      color="purple"
      data-for="vocabChipTip"
      data-place="bottom"
      data-tip="
      <ul>
        <li>
          <span>EN</span>
          <span>Young Boy</span>
        <li>
          <span>JA </span>
          <span lang=\\"ja\\">しょうねん</span>
        </li>
        <li>
          <span>RC</span>
          <span>62%</span>
        </li>
      </ul>
    "
      id={280}
    />
  </styled.ul>
</div>
`;
